library/std/src/sys/unix/fs.rs

```diff
@@ -640,7 +640,10 @@ impl Iterator for ReadDir {
 impl Drop for Dir {
     fn drop(&mut self) {
         let r = unsafe { libc::closedir(self.0) };
-        debug_assert_eq!(r, 0);
+        // a close_dir operation can fail also with EINTR error code
+        // that means that the operation was interrupted by a signal,
+        // and this can be an accepted case.
+        assert!(r == 0 || r == libc::EINTR, "unexpected error during closedir: {:?}", r);
```

Comment:
We might want to use `std::io::Error::last_os_error` here (and elsewhere) if the assert fails, since according to the docs, `closedir` just returns `-1` on failure, which doesn't explain *what* happened.

Reply:
I'm not able to catch your suggestion elsewhere and why to use _probably_ `std::io::Error::last_os_error` can you give a motivation that can convince me?

Reply:
Well, we're showing `r` to the user here. But according to the `closedir` manual, it will only ever return either 0 or -1. So if this assert fails, it will never be anything other than -1, as far as I can see?

> The closedir() function returns 0 on success.  On error, -1 is returned, and errno is set to indicate the error.

(https://linux.die.net/man/3/closedir) (And EINTR doesn't *seem* to be a possible return, especially since that's presumably returned in `errno`, not directly)

Reply:
This should contain the complete list https://pubs.opengroup.org/onlinepubs/007904975/functions/closedir.html 

But I'm still unable to catch why `std::io::Error::last_os_error` and also this will be an API breaking change no?

Reply:
Oh, i'm not saying to return the error as a Result, i'm saying to panic with that error, instead of panicking with `r`. Because closedir will only ever return -1 or 0, it can never return EINTR. That goes in errno.

So if you want to ignore EINTR, you need to check errno anyways.

Reply:
we are interested in `r` in this case because we want to know why we are failing, and maybe different Unix systems can return different `r` value and we want to know this!

Reply:
No, @5225225 is right, the error code is in `errno` not `r`:

> ```
> RETURN VALUE
>        Upon successful completion, closedir() shall return 0; otherwise, -1 shall be returned
>        and errno set to indicate the error.
> ```

Reply:
While waiting for the performance resolution I will be back on this point!

I think I misunderstood the review from the beginning, and my mind was convinced that `r` was the last error.

Now each time that we have a failure we should do something like

```rust
assert_eq!(r, 0, "unexpected error: {:?}", std::io::Error::last_os_error());
```

let's see what the performance result is and we will try to run another time with these changes!

Apologize for the misunderstood here @5225225

Reply:
Panicking in Drop is a bit dodgy, since if we're already panicking, it will abort.

@saethlin pointed out on discord that we could fail to close the directory for a decently sized number of reasons (the fd was closed already, IO errors especially on a network filesystem, ...), none of which are such flaws as to justify aborting the process, and elsewhere we have precedent that dropping `File` ignores all errors.

Reply:
I'm not able to access the `rustc-dev` on discord, so I can not follow the discussion, but I like to keep the same behavior across all the API.

I'm having a hard time to understand when we need to panic or skip panic, looks like a local choice for each call, but I think it is possible to scratch some high-level rules?

Like:
- "Ignore all errors where there are different error causes" like the one in the `File`;
- "using assert instead of debug_assert if os `error` is involved";
- "using debug_assert when the sys call returns a specific error code".

I see this pattern in the review of this PR, but maybe I'm wrong and missing something.

I can have a pointer of the ignore error in the `File`?

Reply:
Using assert in Drop is always dodgy -- it can crash the process. The reason `debug_assert` is permitted is mostly because libstd doesn't ship with debug assertions enabled.

Reply:
So, the option here is:

- continue to use debu_assert and improve the error with a `std::io::Error::last_os_error()`
- skip the check at all?

Reply:
Continue debug_assert and using `std::io::Error::last_os_error()` sounds fine.

Reply:
>Continue debug_assert and using std::io::Error::last_os_error() sounds fine.

Done, not sure if my tentative to avoid evaluating std::io::Error::last_os_error() always is a good thing, but happy to receive feedback on how to improve the code!

If you think that this PR is good enough maybe we can give it a run to see the performance? this change should not impact to at all I think, but not sure